File: C#/34.cs
public class Solution {
    public int[] SearchRange(int[] nums, int target) {
        int left = 0;
        int right = nums.Length - 1;
        int[] result = new int[2];
        result[0] = -1;
        result[1] = -1;
        
        // binary search to find the starting position of target
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target && (mid == 0 || nums[mid-1] != target)) {
                result[0] = mid;
                break;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        // binary search to find the ending position of target
        left = 0;
        right = nums.Length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target && (mid == nums.Length - 1 || nums[mid+1] != target)) {
                result[1] = mid;

File: python/34.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def binarySearch(nums, target, isLeft):
            left, right = 0, len(nums)-1
            while left <= right:
                mid = (left+right)//2
                if nums[mid] == target:
                    if isLeft:
                        if mid == 0 or nums[mid-1] != target:
                            return mid
                        else:
                            right = mid - 1
                    else:
                        if mid == len(nums)-1 or nums[mid+1] != target:
                            return mid
                        else:
                            left = mid + 1
                elif nums[mid] < target:
                    left = mid + 1
                else:
                    right = mid - 1
            return -1
        
        leftIndex = binarySearch(nums, target, True)
        if leftIndex == -1:
            return [-1, -1]
        rightIndex = binarySearch(nums, target, False)
        return [leftIndex, rightIndex]